README
Log-Syslog-Fast version 0.64
============================

This module sends syslog messages over a network socket. It works like
Sys::Syslog in setlogsock's 'udp', 'tcp', or 'unix' modes, but without
the significant CPU overhead of that module when used for high-volume
logging. Use of this specialized module is only recommended if 1) you
must use network syslog as a messaging transport but 2) need to minimize
the time spent in the logger.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Copyright (C) 2009-2011 by Say Media, Inc.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Log::Syslog::Fast.
0.64   Thu Apr  9 13:41:16 PDT 2015
        - Fix PP constructor modifying @_ in place:
          https://github.com/athomason/Log-Syslog-Fast/pull/7
        - Add link to github repo. Both thanks to dolmen@.
0.63   Thu Feb 12 13:50:33 PST 2015
        - Fix missing MANIFEST entries
0.62   Tue Feb 10 17:53:36 PST 2015
        - Emit strict RFC3339 timestamp in RFC5424 mode.
          See https://github.com/athomason/Log-Syslog-Fast/pull/6
          Thanks to Oleg Komarov, komarov@.
0.61   Wed Oct 31 16:41:12 PDT 2012
        - Return a descriptive error on getaddrinfo failure.
0.60   Tue Mar  6 18:00:32 PST 2012
        - Clean up after failed AF_UNIX+SOCK_STREAM socket creation.
          Thanks Pawl Pabian.
0.59   Fri Mar  2 14:11:49 PST 2012
        - Complain instead of crash when string parameters are undefined.
          Thanks Pawl Pabian.
        - Clarify documentation of how $port is used for LOG_UNIX. Also Thanks
          Pawl Pabian.
0.58   Wed Jan  4 11:58:35 PST 2012
        - Add support for messages formatted according to RFC5424. Thanks Ryan
          P. Kelly.
0.57   Thu Dec 29 11:49:30 PST 2011
        - Fix fd leak when reconnecting. Thanks maxs@.
0.56   Sat Nov  5 19:44:36 PDT 2011
        - Fix documentation for set_receiver. Thanks maxs@.
0.55   Mon Apr 11 16:41:43 PDT 2011
        - Recover some speed lost in 0.54 by replacing strcpy with strncpy
        - Add tools for benchmarking different things
        - Replace redundant POD in PP.pm with a pointer to Fast.pm
        - Remove overflow tests which make NetBSD unhappy
0.54   Wed Apr  6 18:20:45 PDT 2011
        - Avoid recomputing length of message string
        - Add ::PP pure-perl version for comparison
        - Make string storage dynamic and growable to limits of memory
0.53   Sat Apr  2 00:47:33 PDT 2011
        - Fix off-by-one buffer overflow (thanks jrockway)
        - Document message length limitation
0.52   Thu Feb 24 18:10:55 PST 2011
        - Move use_ok to separate compilation test
0.51   Wed Feb 23 13:47:02 PST 2011
        - Fix for older NetBSD and OpenBSD which don't define AI_ADDRCONFIG
0.50   Mon Feb 21 15:13:24 PST 2011
        - Fix segfault in BSDs freeaddrinfo when parameter is null
0.49   Fri Feb 18 14:46:07 PST 2011
        - Support IPv6 (courtesy David Skoll)
0.48   Thu Jan 27 17:15:44 PST 2011
        - Croak instead of die when nonexistent methods are called
0.47   Thu Jan 27 17:07:42 PST 2011
        - Add individual setters for facility and severity
        - Add getters for priority, facility, severity, sender, name, and pid
0.46   Tue Nov 16 19:52:24 PST 2010
        - Remove MSG_DONTWAIT from send parameters since appropriate sockets
          (UDP) are already marked non-blocking (thanks POWERMAN@)
        - Add note about thrown exceptions in send
        - croak on close(2) failure
0.45   Mon Sep  6 19:12:59 PDT 2010
        - Report old IO::Socket::INET versions
0.44   Tue Aug 31 04:16:34 PDT 2010
        - Restore legacy aliases
        - Guard against and test for obscure buffer overflows
        - Prevent buffer overflow from big unix filename
0.43   Sun Aug 29 16:42:02 PDT 2010
        - Be more forgiving of time race conditions
0.42   Fri Aug 27 17:44:04 PDT 2010
        - Support perl 5.6.2
0.41   Fri Aug 27 16:55:37 PDT 2010
        - Make pedantic C compilers happier
        - Fix timing race in test
0.40   Thu Aug 26 17:31:13 PDT 2010
        - Eliminate performance penalty of C version by removing perl shim
          layer
        - Rename some things in C space
0.37_C Mon Aug 23 18:13:37 PDT 2010
        - Convert C++ objects to pure C functions; additional perl function
          wrappers cause ~15% performance decrease (this should help with all
          the failures on BSD platforms)
0.37  Sun Aug 22 18:21:11 PDT 2010
        - BSDs appear to raise SIGPIPE on TCP close instead of failing sendto
          syscall, so expect either failure in test
0.36  Sat Aug 21 16:49:20 PDT 2010
        - Don't test against OS- and locale-specific strerror(3) strings
0.35  Fri Aug 20 17:40:31 PDT 2010
        - Overhaul tests to be more readable and comprehensive
        - Support fallback to SOCK_DGRAM with LOG_UNIX should SOCK_STREAM fail
          (thanks to tim.esselens)
        - Add additional tests and docs for LOG_UNIX
0.34  Mon Aug 16 18:53:09 PDT 2010
        - Test that unreachable server fails on connect for TCP/UNIX, or 2nd
          connect for UDP
        - Document that behavior (thanks to Adam Kennedy)
0.33  Wed Jun  9 18:08:47 PDT 2010
        - Bail early if g++ is missing
0.32  Thu May 27 16:37:28 PDT 2010
        - Extra error checking
0.31_01 Mon May 24 14:38:33 PDT 2010
        - Bump version to push test verbosity change to cpan
0.31  Sun May 23 15:07:04 PDT 2010
        - Compute socket(2) length parameter correctly for BSDs
0.30  Sat May 22 17:57:39 PDT 2010
        - s/AF_FILE/AF_UNIX/ to make FreeBSD happy
0.29  Mon May 17 16:19:45 PDT 2010
        - Additional testing of payload
0.28  Mon Jan 11 13:15:10 PST 2010
        - Move syslogconstants into Log::Syslog::Constants
0.27  Fri Jan  8 18:03:41 PST 2010
        - Support unix sockets (e.g. /dev/log)
        - Throw, don't crash on failure in setReceiver
        - Better error reporting
0.26  Fri Jan  8 13:08:43 PST 2010
        - Add get_facility/get_severity for getting constants by name
        - Fix swapped :facilities/:severities
0.25  Thu Nov 19 12:40:03 PST 2009
        - Make log socket be close-on-exec
0.24  Thu Oct 15 22:20:07 PDT 2009
        - Fix porting bug in Log::Syslog::Fast::Simple;
          https://rt.cpan.org/Ticket/Display.html?id=50548
0.23  Thu Jul  2 16:23:09 PST 2009
        - Add shipit config
0.22  Thu Jul  2 16:03:33 PST 2009
        - Clarify newline (non)-handling
0.21  Thu Jul  2 12:00:35 PST 2009
        - Document emit alias
0.20  Tue Jun 30 19:26:33 PST 2009

LogSyslogFast.c
#include "LogSyslogFast.h"

#include <errno.h>
#include <fcntl.h>
#include <netdb.h>
#include <netinet/in.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/un.h>
#include <unistd.h>

#define INITIAL_BUFSIZE 2048

static
void
update_prefix(LogSyslogFast* logger, time_t t)
{
    if (!logger->sender || !logger->name)
        return; /* still initializing */

    logger->last_time = t;

    /* LOG_RFC3164 time string tops out at 15 chars, LOG_RFC5424 at 24 */
    char timestr[25];
    strftime(timestr, 25, logger->time_format, localtime(&t));

    logger->prefix_len = snprintf(
        logger->linebuf, logger->bufsize, logger->msg_format,
        logger->priority, timestr, logger->sender, logger->name, logger->pid
    );

    if (logger->prefix_len > logger->bufsize - 1)
        logger->prefix_len = logger->bufsize - 1;

    /* cache the location in linebuf where msg should be pasted in */
    logger->msg_start = logger->linebuf + logger->prefix_len;
}

LogSyslogFast*
LSF_alloc()
{
    return malloc(sizeof(LogSyslogFast));
}

int
LSF_init(
    LogSyslogFast* logger, int proto, const char* hostname, int port,
    int facility, int severity, const char* sender, const char* name)
{
    if (!logger)
        return -1;

    logger->sock = -1;

    logger->pid = getpid();

    logger->linebuf = malloc(logger->bufsize = INITIAL_BUFSIZE);
    if (!logger->linebuf) {
        logger->err = strerror(errno);
        return -1;
    }

    logger->sender = NULL;
    logger->name = NULL;
    LSF_set_format(logger, LOG_RFC3164);
    LSF_set_sender(logger, sender);
    LSF_set_name(logger, name);

    logger->priority = (facility << 3) | severity;
    update_prefix(logger, time(0));

    return LSF_set_receiver(logger, proto, hostname, port);
}

int
LSF_destroy(LogSyslogFast* logger)
{
    int ret = close(logger->sock);
    if (ret)
        logger->err = strerror(errno);
    free(logger->sender);
    free(logger->name);
    free(logger->linebuf);
    free(logger);
    return ret;
}

void
LSF_set_priority(LogSyslogFast* logger, int facility, int severity)
{
    logger->priority = (facility << 3) | severity;
    update_prefix(logger, time(0));
}

void
LSF_set_facility(LogSyslogFast* logger, int facility)
{
    LSF_set_priority(logger, facility, LSF_get_severity(logger));
}

void
LSF_set_severity(LogSyslogFast* logger, int severity)
{
    LSF_set_priority(logger, LSF_get_facility(logger), severity);
}

int
LSF_set_sender(LogSyslogFast* logger, const char* sender)
{
    free(logger->sender);
    logger->sender = strdup(sender);
    if (!logger->sender) {
        logger->err = "strdup failure in set_sender";
        return -1;
    }
    update_prefix(logger, time(0));
    return 0;
}

int
LSF_set_name(LogSyslogFast* logger, const char* name)
{
    free(logger->name);
    logger->name = strdup(name);
    if (!logger->name) {
        logger->err = "strdup failure in set_name";
        return -1;
    }
    update_prefix(logger, time(0));
    return 0;
}

void
LSF_set_pid(LogSyslogFast* logger, int pid)
{
    logger->pid = pid;
    update_prefix(logger, time(0));
}

int
LSF_set_format(LogSyslogFast* logger, int format)
{
    logger->format = format;

    /* msg_format must contain format strings for:
       1) priority (int)
       2) timestamp (string)
       3) hostname (string)
       4) app name (string)
       5) pid (int)
    */
    if (logger->format == LOG_RFC3164) {
        /*
           'The TIMESTAMP field is the local time and is in the format of
           "Mmm dd hh:mm:ss"'

            Example: "Jan  4 11:22:33"
        */
        logger->time_format = "%h %e %H:%M:%S";
        logger->msg_format = "<%d>%s %s %s[%d]: ";
    }
    else if (logger->format == LOG_RFC5424) {
        /*
            TIMESTAMP       = NILVALUE / FULL-DATE "T" FULL-TIME
            FULL-DATE       = DATE-FULLYEAR "-" DATE-MONTH "-" DATE-MDAY
            DATE-FULLYEAR   = 4DIGIT
            DATE-MONTH      = 2DIGIT  ; 01-12
            DATE-MDAY       = 2DIGIT  ; 01-28, 01-29, 01-30, 01-31 based on
                                        ; month/year
            FULL-TIME       = PARTIAL-TIME TIME-OFFSET
            PARTIAL-TIME    = TIME-HOUR ":" TIME-MINUTE ":" TIME-SECOND
                                [TIME-SECFRAC]
            TIME-HOUR       = 2DIGIT  ; 00-23
            TIME-MINUTE     = 2DIGIT  ; 00-59
            TIME-SECOND     = 2DIGIT  ; 00-59
            TIME-SECFRAC    = "." 1*6DIGIT
            TIME-OFFSET     = "Z" / TIME-NUMOFFSET
            TIME-NUMOFFSET  = ("+" / "-") TIME-HOUR ":" TIME-MINUTE

            Example: "2012-01-04T11:22:33-0800"
        */
        logger->time_format = "%Y-%m-%dT%H:%M:%S%z";

        /* STRUCTURED-DATA and MSGID fields are omitted */
        logger->msg_format = "<%d>1 %s %s %s %d - - ";
    }
    else {
        logger->err = "invalid format constant";
        return -1;
    }
    update_prefix(logger, time(0));
    return 0;
}

#ifdef AF_INET6
#define clean_return(x) if (results) freeaddrinfo(results); return x;
#else
#define clean_return(x) return x;
#endif

/* must match constants in LogSyslogFast.pm */
#define LOG_UDP  0
#define LOG_TCP  1
#define LOG_UNIX 2

int
LSF_set_receiver(LogSyslogFast* logger, int proto, const char* hostname, int port)
{
    const struct sockaddr* p_address;
    int address_len;
#ifdef AF_INET6
    struct addrinfo* results = NULL;
#endif

    if (logger->sock >= 0) {
        int ret = close(logger->sock);
        if (ret) {
            logger->err = strerror(errno);
            return -1;
        }
    }

    /* set up a socket, letting kernel assign local port */
    if (proto == LOG_UDP || proto == LOG_TCP) {

#ifdef AF_INET6

/* For NetBSD: http://www.mail-archive.com/bug-gnulib@gnu.org/msg17067.html */
#ifndef AI_ADDRCONFIG
#define AI_ADDRCONFIG 0
#endif

/* For MacOS: http://mailman.videolan.org/pipermail/vlc-devel/2008-May/044005.html */
#ifndef AI_NUMERICSERV
#define AI_NUMERICSERV 0
#endif

        struct addrinfo *rp;
        struct addrinfo hints;
        char portstr[32];
        int r;

        snprintf(portstr, sizeof(portstr), "%d", port);
        memset(&hints, 0, sizeof(hints));
        hints.ai_flags = AI_ADDRCONFIG | AI_NUMERICSERV;
        hints.ai_family = AF_UNSPEC;
        if (proto == LOG_TCP) {
            hints.ai_socktype = SOCK_STREAM;
        } else {
            hints.ai_socktype = SOCK_DGRAM;
        }
        hints.ai_protocol = 0;
        hints.ai_addrlen = 0;
        hints.ai_addr = NULL;
        hints.ai_canonname = NULL;
        hints.ai_next = NULL;

        r = getaddrinfo(hostname, portstr, &hints, &results);
        if (r < 0) {
            logger->err = gai_strerror(r);
            return -1;
        }
        else if (!results) {
            logger->err = "no results from getaddrinfo";
            return -1;
        }
        for (rp = results; rp != NULL; rp = rp->ai_next) {
            logger->sock = socket(rp->ai_family, rp->ai_socktype, 0);
            if (logger->sock == -1) {
                r = errno;
                continue;
            }
            p_address = rp->ai_addr;
            address_len = rp->ai_addrlen;
            break;
        }
        if (logger->sock == -1) {
            logger->err = "socket failure";
            clean_return(-1);
        }

#else /* !AF_INET6 */

        /* resolve the remote host */
        struct hostent* host = gethostbyname(hostname);
        if (!host || !host->h_addr_list || !host->h_addr_list[0]) {
            logger->err = "resolve failure";
            return -1;
        }

        /* create the remote host's address */
        struct sockaddr_in raddress;
        raddress.sin_family = AF_INET;
        memcpy(&raddress.sin_addr, host->h_addr_list[0], sizeof(raddress.sin_addr));
        raddress.sin_port = htons(port);
        p_address = (const struct sockaddr*) &raddress;
        address_len = sizeof(raddress);

        /* construct socket */
        if (proto == LOG_UDP) {
            logger->sock = socket(AF_INET, SOCK_DGRAM, 0);

            /* make the socket non-blocking */
            int flags = fcntl(logger->sock, F_GETFL, 0);
            fcntl(logger->sock, F_SETFL, flags | O_NONBLOCK);
            flags = fcntl(logger->sock, F_GETFL, 0);
            if (!(flags & O_NONBLOCK)) {
                logger->err = "nonblock failure";
                return -1;
            }
        }
        else if (proto == LOG_TCP) {
            logger->sock = socket(AF_INET, SOCK_STREAM, 0);
        }

#endif /* AF_INET6 */
    }
    else if (proto == LOG_UNIX) {

        /* create the log device's address */
        struct sockaddr_un raddress;
        raddress.sun_family = AF_UNIX;
        strncpy(raddress.sun_path, hostname, sizeof(raddress.sun_path) - 1);
        p_address = (const struct sockaddr*) &raddress;
        address_len = sizeof(raddress);

        /* construct socket */
        logger->sock = socket(AF_UNIX, SOCK_STREAM, 0);
    }
    else {
        logger->err = "bad protocol";
        return -1;
    }

    if (logger->sock < 0) {
        logger->err = strerror(errno);
        clean_return(-1);
    }

    /* close the socket after exec to match normal Perl behavior for sockets */
    fcntl(logger->sock, F_SETFD, FD_CLOEXEC);

    /* connect the socket */
    if (connect(logger->sock, p_address, address_len) != 0) {
        /* some servers (rsyslog) may use SOCK_DGRAM for unix domain sockets */
        if (proto == LOG_UNIX && errno == EPROTOTYPE) {
            /* clean up existing bad socket */
            close(logger->sock);
            if (logger->sock < 0) {
                logger->err = strerror(errno);
                clean_return(-1);
            }

            logger->sock = socket(AF_UNIX, SOCK_DGRAM, 0);
            if (connect(logger->sock, p_address, address_len) != 0) {
                logger->err = strerror(errno);
                clean_return(-1);
            }
        }
        else {
            logger->err = strerror(errno);
            clean_return(-1);
        }
    }

    clean_return(0);
}

int
LSF_send(LogSyslogFast* logger, const char* msg_str, int msg_len, time_t t)
{
    /* update the prefix if seconds have rolled over */
    if (t != logger->last_time)
        update_prefix(logger, t);

    int line_len = logger->prefix_len + msg_len;
    /* ensure there's space in the buffer for total length including a trailing NULL */
    if (logger->bufsize < line_len + 1) {
        /* try to increase buffer */
        int new_bufsize = 2 * logger->bufsize;
        while (new_bufsize < line_len + 1)
            new_bufsize *= 2;
        if (new_bufsize < 0) {
            /* overflow */
            logger->err = "message too large";
            return -1;
        }

        char* new_buf = realloc(logger->linebuf, new_bufsize);
        if (!new_buf) {
            logger->err = strerror(errno);
            return -1;
        }

        logger->linebuf = new_buf;
        logger->bufsize = new_bufsize;
        logger->msg_start = logger->linebuf + logger->prefix_len;
    }

    /* paste the message into linebuf just past where the prefix was placed */
    memcpy(logger->msg_start, msg_str, msg_len + 1); /* include perl-added null */

    int ret = send(logger->sock, logger->linebuf, line_len, 0);

    if (ret < 0)
        logger->err = strerror(errno);
    return ret;
}

int
LSF_get_priority(LogSyslogFast* logger)
{
    return logger->priority;
}

int
LSF_get_facility(LogSyslogFast* logger)
{
    return logger->priority >> 3;
}

int
LSF_get_severity(LogSyslogFast* logger)
{
    return logger->priority & 7;
}

const char*
LSF_get_sender(LogSyslogFast* logger)
{
    return logger->sender;
}

const char*
LSF_get_name(LogSyslogFast* logger)
{
    return logger->name;
}

int
LSF_get_pid(LogSyslogFast* logger)
{
    return logger->pid;
}

int
LSF_get_sock(LogSyslogFast* logger)
{
    return logger->sock;
}

int
LSF_get_format(LogSyslogFast* logger)
{
    return logger->format;
}
